### DISCUSSION TITLE
Unsupported Zigbee Device

### DISCUSSION THREAD

**Discussion by Blue-beef** [Koenkk/zigbee2mqtt#16397] [Devices]
I have successfully added a Salus ECM600 Electric Meter, which has four CT clamps measuring instantaneous power. I used the template and examples from other devices to create an external converter file. I do not fully understand in the inner workings of zigbee, but with trial and error managed to expose Link Quality and the first CT clamp power measurement.

However, I cannot figure out a way to expose the other three power channels.

Here is my external converter file:

const fz = require(‘zigbee-herdsman-converters/converters/fromZigbee’);
const tz = require(‘zigbee-herdsman-converters/converters/toZigbee’);
const exposes = require(‘zigbee-herdsman-converters/lib/exposes’);
const reporting = require(‘zigbee-herdsman-converters/lib/reporting’);
const extend = require(‘zigbee-herdsman-converters/lib/extend’);
const ota = require(‘zigbee-herdsman-converters/lib/ota’);
const e = exposes.presets;
const ea = exposes.access;

const definition = {
zigbeeModel: [‘ECM600’],
model: ‘ECM600’, // Vendor model number, look on the device for a model number
vendor: ‘Salus Controls’, // Vendor of the device (only used for documentation and startup logging)
description: ‘Electric Monitor’, // Description of the device, copy from vendor site. (only used for documentation and startup logging)
fromZigbee: [fz.metering],
toZigbee: , // Should be empty, unless device can be controlled (e.g. lights, switches).
exposes: [e.power()], // Defines what this device exposes, used for e.g. Home Assistant discovery and in the frontend
configure: async (device, coordinatorEndpoint, logger) => {
const endpoint = device.getEndpoint(1);
await reporting.bind(endpoint, coordinatorEndpoint, [‘seMetering’]);
},
ota: ota.salus,
};

module.exports = definition;

Any help would be very much appreciated.

**Comment by Great-Chart** [comment]
Can I ask if you managed to get any further with this?
I ordered one of these to explore but the above converter threw up an error message (on the exposes line) that I struggled with and before I get too far into confusing myself I thought I'd see if you got any further.

**Reply by Blue-beef** [reply]
Unfortunately, no I haven't. I would be keen to know whether succeed.

**Reply by Great-Chart** [reply]
I'm struggling with indentation / annotation at present (some show as errors in the code / some appear fine; but stops Z2M running.
But I'll keep trying to work through that (from examples).
If you can copy the code for your converter (above) that presents with the correct annotations that would likely help me get past this stumbling block at least

**Reply by Blue-beef** [reply]
I've attached the current config file I'm using.
[salusecm600.js.txt](https://github.com/Koenkk/zigbee2mqtt/files/12560450/salusecm600.js.txt)

**Reply by Great-Chart** [reply]
I'd missed out the line below and thus wherever I'd inserted the "ota: ota.salus," line it'd stop Z2M running.
_const ota = require('zigbee-herdsman-converters/lib/ota');_

So I've got as far as trying to update old firmware (20160106) and presently I've moved from power showing N/A to now showing as Null.
Like you I just get one reading and as I'm sure you've found out we're not able to directly trigger this device to see what's incoming to Z2M via logs.

I'm still slightly behind you so I suspect I'll need to try various CT clamps on various feeds to at least get one working.
For me it's just on a single phase supply but the (limited) documentation indicates incomer + "3-others" so I guess we just keep trying various things.
Now I've got this far I feel I'll be able to make a bit more sense of the existing supported converters.
So thanks for getting me pointed in the right direction - did you get OTA updates to work?

**Reply by Blue-beef** [reply]
No, I didn't want to risk the update!

**Comment by Great-Chart** [comment]
OK I've had a bit of a dabble with Z2M Dev console and made some sense (starting from a low bar mind)!
This device presents 4-off end points inferring that CT clamps 2 thru 4 would be endpoints 2 thru 4.  If we use Dev console set to a differing end point, Cluster selecting seMetering and Attributes for "instantaneousDemand" and "currSummDelivered" then the values for energy & power will change to that reported by the other CT clamps (end-points).
But I've not fathomed out how to report and see them all.

I've pulled in some content from Salus SP600 (smart plug with monitoring) and pulled battery data from other sources to at least get things a little more populated.
Refer below:-

Exposes within Z2M now lists Battery, Energy, Power and Linkquality
Bind within Z2M now shows data for source endpoints 1 thru 4
Reporting within Z2M is limited to endpoint #1 with instantaneousDemand, currentSummDelivered and batteryVoltage (all as per SP600)
State within Z2m lists battery, battery low, energy, linkquality, power, update, update available and voltage

But not really that close to getting CT's 2 thru 4 reading; but there's a few lines commented out (below) that MAY likely be on the right track???

`const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['ECM600'], // The model ID from: Device with modelID 'lumi.sens' is not supported.
    model: 'ECM600', // Vendor model number, look on the device for a model number
    vendor: 'Salus Controls', // Vendor of the device (only used for documentation and startup logging)
    description: 'Electricity Monitor', // Description of the device, copy from vendor site. (only used for documentation and startup logging)
    fromZigbee: [fz.metering, fz.battery], // We will add this later
    toZigbee: [], // Should be empty, unless device can be controlled (e.g. lights, switches).
    exposes: [e.battery(), e.energy(), e.power()], // Defines what this device exposes, used for e.g. Home Assistant discovery and in the frontend
//    meta: {multiEndpoint: true}, // Noted this line within EXAMPLE Tuya converters; likely relevant
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(1);
// NB - Seems possible to change endpoint# and get differing energy / power readings - BUT just one of them gets reported        
        await reporting.bind(endpoint, coordinatorEndpoint, ['seMetering', 'genPowerCfg']);
        await reporting.instantaneousDemand(endpoint, {min: 5, change: 10});
        await reporting.currentSummDelivered(endpoint, {min: 5, change: [0, 10]});
        await endpoint.read('seMetering', ['multiplier', 'divisor']);       
        await reporting.batteryVoltage(endpoint);
},
    ota: ota.salus,
};

module.exports = definition;`

This has gone through multiple iterations of putting things in to trail and taking things out I'm sure it's now very messy but it presents the info listed above and woudl appear to be largely fully functional for this device IF it was just a single channel (endpoint) unit; but of course there are 3 more CT's available inferring power & energy for those can still be unlocked with more trial & error (or some more knowledgeable direction).
